The ```pip``` container seems to be missing ```/usr/bin/koji``` according to this [test failure](https://ci.centos.org/job/bodhi-pipeline/job/PR-2837/61/testReport/devel.ci.integration.tests/test_bodhi_cli/test_updates_download/):

```
bodhi_container = DockerContainer(image=bodhi-ci-integration-bodhi/pip:latest, id=ed7e2e1bb601ea02719e66f97006ff31ef68aa3b7be2bd0fa2f4e926e8e1964e)
db_container = DockerContainer(image=postgres:latest, id=2a3bf697f4e32e71c1233b1a9b1b5c4a50769bc6d15cd96cc9ae91e946cbc30c)

    def test_updates_download(bodhi_container, db_container):
        """Test ``bodhi updates download``"""
        # Fetch the last updates from the DB
        builds = []
        db_ip = db_container.get_IPv4s()[0]
        query_updates = "SELECT id, alias, title FROM updates ORDER BY date_submitted DESC LIMIT 3"
        query_builds = "SELECT nvr FROM builds WHERE update_id = %s ORDER BY nvr"
        conn = psycopg2.connect("dbname=bodhi2 user=postgres host={}".format(db_ip))
        with conn:
            with conn.cursor() as curs:
                curs.execute(query_updates)
                updates = [_db_record_to_munch(curs, record) for record in curs]
                for update in updates:
                    curs.execute(query_builds, (update.id, ))
                    builds.extend([r[0] for r in curs])
        conn.close()
        # Prepare the command to run
        cmd = [
            "updates", "download", "--arch", "all",
            "--updateid", ",".join([u.alias for u in updates]),
        ]
        # The bodhi CLI will execute the koji CLI. Replace that executable with
        # something we can track.
        koji_mock = "#!/bin/sh\necho TESTING CALL $0 $@\n"
>       with replace_file(bodhi_container, "/usr/bin/koji", koji_mock):

devel/ci/integration/tests/test_bodhi_cli.py:317: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib64/python3.6/contextlib.py:81: in __enter__
    return next(self.gen)
devel/ci/integration/tests/utils.py:137: in replace_file
    container.copy_from(path, backupfile)
/usr/local/lib/python3.6/site-packages/conu/backend/docker/container.py:449: in copy_from
    run_cmd(cmd)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cmd = ['docker', 'cp', 'ed7e2e1bb601ea02719e66f97006ff31ef68aa3b7be2bd0fa2f4e926e8e1964e:/usr/bin/koji', '/tmp/tmpr6_gvkdv/original']
return_output = False, ignore_status = False, kwargs = {}
process = <subprocess.Popen object at 0x7f690e6ea048>
output = 'Error response from daemon: lstat /var/lib/docker/overlay2/e689e12994519b239f0241918e2c53e41977f930ba199d22e8aef12ce0844a22/merged/usr/bin/koji: no such file or directory\n'

    def run_cmd(cmd, return_output=False, ignore_status=False, **kwargs):
        """
        run provided command on host system using the same user as you invoked this code, raises
        subprocess.CalledProcessError if it fails
    
        :param cmd: list of str
        :param return_output: bool, return output of the command
        :param ignore_status: bool, do not fail in case nonzero return code
        :param kwargs: pass keyword arguments to subprocess.check_* functions; for more info,
                please check `help(subprocess.Popen)`
        :return: None or str
        """
        logger.debug('command: "%s"' % ' '.join(cmd))
        process = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT,
                                   universal_newlines=True, **kwargs)
        output = process.communicate()[0]
        logger.debug(output)
    
        if process.returncode > 0:
            if ignore_status:
                if return_output:
                    return output
                else:
                    return process.returncode
            else:
>               raise subprocess.CalledProcessError(cmd=cmd, returncode=process.returncode)
E               subprocess.CalledProcessError: Command '['docker', 'cp', 'ed7e2e1bb601ea02719e66f97006ff31ef68aa3b7be2bd0fa2f4e926e8e1964e:/usr/bin/koji', '/tmp/tmpr6_gvkdv/original']' returned non-zero exit status 1.

/usr/local/lib/python3.6/site-packages/conu/utils/__init__.py:137: CalledProcessError
```

--- a/.mergify.yml
+++ b/.mergify.yml
@@ -14,12 +14,14 @@ pull_request_rules:
   - status-success=f28-python2-unit
   - status-success=f28-python3-unit
   - status-success=f28-python3-diff-cover
+  - status-success=f28-python3-integration
   - status-success=f29-docs
   - status-success=f29-flake8
   - status-success=f29-pydocstyle
   - status-success=f29-python2-unit
   - status-success=f29-python3-unit
   - status-success=f29-python3-diff-cover
+  - status-success=f29-python3-integration
   - status-success=pip-docs
   - status-success=pip-flake8
   - status-success=pip-mypy
@@ -27,7 +29,7 @@ pull_request_rules:
   - status-success=pip-python2-unit
   - status-success=pip-python3-unit
   - status-success=pip-python3-diff-cover
-  - status-success=f29-python3-integration
+  - status-success=pip-python3-integration
   name: default
 - actions:
     merge:
@@ -45,12 +47,14 @@ pull_request_rules:
   - status-success=f28-python2-unit
   - status-success=f28-python3-unit
   - status-success=f28-python3-diff-cover
+  - status-success=f28-python3-integration
   - status-success=f29-docs
   - status-success=f29-flake8
   - status-success=f29-pydocstyle
   - status-success=f29-python2-unit
   - status-success=f29-python3-unit
   - status-success=f29-python3-diff-cover
+  - status-success=f29-python3-integration
   - status-success=pip-docs
   - status-success=pip-flake8
   - status-success=pip-mypy
@@ -58,6 +62,7 @@ pull_request_rules:
   - status-success=pip-python2-unit
   - status-success=pip-python3-unit
   - status-success=pip-python3-diff-cover
+  - status-success=pip-python3-integration
   name: default-from-bowlofeggs
 - actions:
     backport:

--- a/devel/ci/Dockerfile-f28
+++ b/devel/ci/Dockerfile-f28
@@ -25,27 +25,33 @@ RUN dnf install --disablerepo rawhide-modular -y \
     python3-arrow \
     python3-bleach \
     python3-bugzilla \
+    python3-click \
     python3-colander \
+    python3-cryptography \
     python3-diff-cover \
     python3-dogpile-cache \
     python3-fedmsg \
     python3-fedora \
     python3-feedgen \
     python3-flake8 \
+    python3-jinja2 \
     python3-kitchen \
     python3-libcomps \
     python3-librepo \
     python3-markdown \
     python3-mock \
     python3-munch \
+    python3-openid \
     python3-pillow \
     python3-pydocstyle \
     python3-pylibravatar \
+    python3-pyramid \
     python3-pyramid-mako \
     python3-pytest \
     python3-pytest-cov \
     python3-responses \
     python3-simplemediawiki \
+    python3-six \
     python3-sqlalchemy \
     python3-webtest \
     python2-munch \

--- a/devel/ci/Dockerfile-f29
+++ b/devel/ci/Dockerfile-f29
@@ -26,27 +26,33 @@ RUN dnf install --disablerepo rawhide-modular -y \
     python3-alembic \
     python3-arrow \
     python3-bleach \
+    python3-click \
     python3-colander \
+    python3-cryptography \
     python3-diff-cover \
     python3-dogpile-cache \
     python3-fedmsg \
     python3-fedora \
     python3-feedgen \
     python3-flake8 \
+    python3-jinja2 \
     python3-kitchen \
     python3-libcomps \
     python3-librepo \
     python3-markdown \
     python3-mock \
     python3-munch \
+    python3-openid \
     python3-pillow \
     python3-pydocstyle \
     python3-pylibravatar \
+    python3-pyramid \
     python3-pyramid-mako \
     python3-pytest \
     python3-pytest-cov \
     python3-responses \
     python3-simplemediawiki \
+    python3-six \
     python3-sqlalchemy \
     python3-webtest \
     python2-munch \

--- a/devel/ci/Dockerfile-rawhide
+++ b/devel/ci/Dockerfile-rawhide
@@ -8,21 +8,27 @@ RUN dnf install --disablerepo rawhide-modular -y \
     findutils \
     git \
     liberation-mono-fonts \
+    python3-dnf \
+    python3-yaml \
     make \
     python2-click \
     python2-fedora \
     python2-koji \
     python2-mock \
     python2-munch \
+    python2-pytest \
     python2-pytest-cov \
+    python2-six \
     python3-alembic \
     python3-arrow \
     python3-bleach \
     python3-bugzilla \
+    python3-click \
     python3-colander \
     python3-cornice \
     python3-cornice-sphinx \
     python3-createrepo_c \
+    python3-cryptography \
     python3-diff-cover \
     python3-dogpile-cache \
     python3-fedmsg \
@@ -30,6 +36,7 @@ RUN dnf install --disablerepo rawhide-modular -y \
     python3-feedgen \
     python3-flake8 \
     python3-hawkey \
+    python3-jinja2 \
     python3-kitchen \
     python3-koji \
     python3-libcomps \
@@ -37,17 +44,21 @@ RUN dnf install --disablerepo rawhide-modular -y \
     python3-markdown \
     python3-mock \
     python3-munch \
+    python3-openid \
     python3-pillow \
     python3-pydocstyle \
     python3-pylibravatar \
+    python3-pyramid \
     python3-pyramid-fas-openid \
     python3-pyramid-mako \
     python3-pytest \
     python3-pytest-cov \
     python3-responses \
     python3-simplemediawiki \
+    python3-six \
     python3-sqlalchemy \
     python3-sqlalchemy_schemadisplay \
+    python3-waitress \
     python3-webtest
 
 # Fake pungi being installed so we can avoid it and all its dependencies

--- a/devel/ci/bodhi-ci
+++ b/devel/ci/bodhi-ci
@@ -182,7 +182,15 @@ def build(concurrency, container_runtime, failfast, release, tty):
 def clean(concurrency, container_runtime, init, release, tty):
     """Remove all builds pertaining to Bodhi CI."""
     buffer_output = concurrency != 1 or len(release) != 1
-    clean_jobs = [CleanJob(r, buffer_output=buffer_output) for r in release]
+    clean_jobs = [
+        CleanJob(r, buffer_output=buffer_output) for r in release
+    ] + [
+        IntegrationCleanJob(app_name="bodhi", release=r, buffer_output=buffer_output)
+        for r in release
+    ] + [
+        IntegrationCleanJob(app_name=app_name, release="prod", buffer_output=buffer_output)
+        for app_name in ("resultsdb", "waiverdb", "greenwave")
+    ]
     _run_jobs(clean_jobs)
 
 
@@ -272,6 +280,18 @@ def unit(archive, concurrency, container_runtime, no_build, failfast, init, pyve
     """Run the unit tests."""
     _run_jobs(_build_jobs_list('unit', concurrency, release, pyvers=pyver, archive=archive,
                                archive_path=archive_path))
+
+
+@cli.command("integration-build")
+@concurrency_option
+@container_runtime_option
+@failfast_option
+@pyver_option
+@releases_option
+@tty_option
+def integration_build(concurrency, container_runtime, failfast, pyver, release, tty):
+    """Build the containers for integration testing."""
+    _run_jobs(_build_jobs_list('integration-build', concurrency, release, pyvers=pyver))
 
 
 @cli.command()
@@ -989,6 +1009,29 @@ class IntegrationJob(Job):
             )
 
 
+class IntegrationCleanJob(Job):
+    """
+    Define a Job for removing all container images built by bodhi-ci.
+
+    See the Job superclass's docblock for details about its attributes.
+    """
+
+    _label = 'integration-clean'
+
+    def __init__(self, app_name, *args, **kwargs):
+        """
+        Initialize the IntegrationCleanJob.
+
+        See the superclass's docblock for details about accepted parameters.
+        """
+        super(IntegrationCleanJob, self).__init__(*args, **kwargs)
+
+        self._container_image = '{}-integration-{}'.format(CONTAINER_NAME, app_name)
+        if app_name == "bodhi":
+            self._container_image = "{}/{}".format(self._container_image, self.release)
+        self._command = [container_runtime, 'rmi', self._container_image]
+
+
 def _build_jobs_list(
         command: str, concurrency: int, releases: typing.Sequence[str],
         pyvers: typing.Sequence[int] = None, archive: bool = False,
@@ -1018,16 +1061,15 @@ def _build_jobs_list(
     else:
         buffer_output = concurrency != 1 and len(releases) != 1
     jobs = []  # type: typing.List[Job]
-    if command in ('all', 'integration') and 3 in pyvers:
+    if command in ('all', 'integration-build', 'integration') and 3 in pyvers:
         integration_deps_build_jobs = [
-            IntegrationBuildJob(app_name="resultsdb", release="prod", buffer_output=buffer_output),
-            IntegrationBuildJob(app_name="waiverdb", release="prod", buffer_output=buffer_output),
-            IntegrationBuildJob(app_name="greenwave", release="prod", buffer_output=buffer_output),
+            IntegrationBuildJob(app_name=app_name, release="prod", buffer_output=buffer_output)
+            for app_name in ("resultsdb", "waiverdb", "greenwave")
         ]
         jobs.extend(integration_deps_build_jobs)
     for release in releases:
         if command in ('all', 'build', 'docs', 'flake8', 'mypy', 'pydocstyle', 'diff_cover',
-                       'unit'):
+                       'unit', 'integration-build', 'integration'):
             # We don't want to build non-pip releases if the command is mypy; mypy only uses pip.
             if command != 'mypy' or release == 'pip':
                 build_job = BuildJob(release, buffer_output=buffer_output)
@@ -1059,11 +1101,12 @@ def _build_jobs_list(
                         archive=archive, archive_path=archive_path, pyver=pyver,
                         release=release, depends_on=unit_job, buffer_output=buffer_output)
                     jobs.append(diff_cover_job)
-        if command in ('all', 'integration') and 3 in pyvers:
+        if command in ('all', 'integration-build', 'integration') and 3 in pyvers:
             integration_build_job = IntegrationBodhiBuildJob(
                 release=release, depends_on=build_job, buffer_output=buffer_output
             )
             jobs.append(integration_build_job)
+        if command in ('all', 'integration') and 3 in pyvers:
             integration_job = IntegrationJob(
                 archive=archive, archive_path=archive_path, pyver=3, release=release,
                 depends_on=integration_deps_build_jobs + [integration_build_job],

--- a/devel/ci/bodhi-ci
+++ b/devel/ci/bodhi-ci
@@ -389,6 +389,9 @@ class Job(object):
             self._popen_kwargs['stderr'] = subprocess.STDOUT
         self._stdout = b''
 
+    def __repr__(self):
+        return "<{} release={!r}>".format(self.__class__.__name__, self.release)
+
     @property
     def label(self):
         """
@@ -907,11 +910,41 @@ class IntegrationBuildJob(BuildJob):
         """
         super(IntegrationBuildJob, self).__init__(*args, **kwargs)
 
+        self._app_name = app_name
         self._label = 'integration-build-{}'.format(app_name)
         self._container_image = '{}-integration-{}'.format(CONTAINER_NAME, app_name)
-        dockerfile = os.path.join(PROJECT_PATH, 'devel', 'ci', 'integration', app_name, 'Dockerfile')
-        self._command = [container_runtime, 'build', '--force-rm', '--pull', '-t',
-                         self._container_image, '-f', dockerfile, '.']
+        dockerfile = os.path.join(
+            PROJECT_PATH, 'devel', 'ci', 'integration', app_name, 'Dockerfile'
+        )
+        self._command = [container_runtime, 'build', '--force-rm', '--pull',
+                         '-t', self._container_image, '-f', dockerfile, '.']
+
+    def __repr__(self):
+        return "<{} app={!r}>".format(self.__class__.__name__, self._app_name)
+
+
+class IntegrationBodhiBuildJob(IntegrationBuildJob):
+    """
+    Build Bodhi in a container image for integration testing.
+
+    See the Job superclass's docblock for details about its attributes.
+    """
+
+    def __init__(self, *args, **kwargs):
+        """
+        Initialize the IntegrationBodhiBuildJob.
+
+        See the superclass's docblock for details about accepted parameters.
+        """
+        super(IntegrationBodhiBuildJob, self).__init__(*args, app_name="bodhi", **kwargs)
+
+        self._container_image = '{}-integration-bodhi/{}'.format(CONTAINER_NAME, self.release)
+        dockerfile = os.path.join(
+            PROJECT_PATH, 'devel', 'ci', 'integration', 'bodhi',
+            'Dockerfile-{}'.format(self.release),
+        )
+        self._command = [container_runtime, 'build', '--force-rm', '-t', self._container_image,
+                         '-f', dockerfile, '.']
 
 
 class IntegrationJob(Job):
@@ -943,15 +976,15 @@ class IntegrationJob(Job):
 
         self.pyver = pyver
         self._command = [sys.executable, '-m', 'pytest', '--no-cov', 'devel/ci/integration/tests/']
-        bodhi_container_image = '{}-integration-bodhi'.format(CONTAINER_NAME)
+        bodhi_container_image = '{}-integration-bodhi/{}'.format(CONTAINER_NAME, self.release)
         self._popen_kwargs["env"] = os.environ.copy()
         self._popen_kwargs["env"]["BODHI_INTEGRATION_IMAGE"] = bodhi_container_image
         if failfast:
             self._command.append('-x')
         if archive:
             self._command.append(
-                '--junit-xml={}/{}/nosetests.xml'.format(
-                    archive_path, self._label
+                '--junit-xml={}/{}-{}/nosetests.xml'.format(
+                    archive_path, self.release, self._label
                 )
             )
 
@@ -985,6 +1018,13 @@ def _build_jobs_list(
     else:
         buffer_output = concurrency != 1 and len(releases) != 1
     jobs = []  # type: typing.List[Job]
+    if command in ('all', 'integration') and 3 in pyvers:
+        integration_deps_build_jobs = [
+            IntegrationBuildJob(app_name="resultsdb", release="prod", buffer_output=buffer_output),
+            IntegrationBuildJob(app_name="waiverdb", release="prod", buffer_output=buffer_output),
+            IntegrationBuildJob(app_name="greenwave", release="prod", buffer_output=buffer_output),
+        ]
+        jobs.extend(integration_deps_build_jobs)
     for release in releases:
         if command in ('all', 'build', 'docs', 'flake8', 'mypy', 'pydocstyle', 'diff_cover',
                        'unit'):
@@ -1019,22 +1059,16 @@ def _build_jobs_list(
                         archive=archive, archive_path=archive_path, pyver=pyver,
                         release=release, depends_on=unit_job, buffer_output=buffer_output)
                     jobs.append(diff_cover_job)
-    if command in ('all', 'integration') and 3 in pyvers and "f29" in releases:
-        build_jobs = [
-            IntegrationBuildJob(app_name="resultsdb", release="f29",
-                                buffer_output=buffer_output),
-            IntegrationBuildJob(app_name="waiverdb", release="f29",
-                                buffer_output=buffer_output),
-            IntegrationBuildJob(app_name="greenwave", release="f29",
-                                buffer_output=buffer_output),
-            IntegrationBuildJob(app_name="bodhi", release="f29",
-                                buffer_output=buffer_output),
-        ]
-        jobs.extend(build_jobs)
-        integration_job = IntegrationJob(
-            archive=archive, archive_path=archive_path, pyver=3, release="f29",
-            depends_on=build_jobs, buffer_output=buffer_output)
-        jobs.append(integration_job)
+        if command in ('all', 'integration') and 3 in pyvers:
+            integration_build_job = IntegrationBodhiBuildJob(
+                release=release, depends_on=build_job, buffer_output=buffer_output
+            )
+            jobs.append(integration_build_job)
+            integration_job = IntegrationJob(
+                archive=archive, archive_path=archive_path, pyver=3, release=release,
+                depends_on=integration_deps_build_jobs + [integration_build_job],
+                buffer_output=buffer_output)
+            jobs.append(integration_job)
 
     return jobs
 

--- a/devel/ci/cico.pipeline
+++ b/devel/ci/cico.pipeline
@@ -120,7 +120,8 @@ def test_release = { String release ->
             bodhi_ci(release, 'diff_cover', 'python3-diff-cover', '--no-build --no-init -p 3')
         },
         integration: {
-            bodhi_ci(release, 'integration', 'python3-integration', '--no-init -p 3')
+            bodhi_ci(release, 'integration-build', 'python3-integration-build', '-p 3')
+            bodhi_ci(release, 'integration', 'python3-integration', '--no-build --no-init -p 3')
         }
     )
 }

--- a/devel/ci/cico.pipeline
+++ b/devel/ci/cico.pipeline
@@ -120,9 +120,7 @@ def test_release = { String release ->
             bodhi_ci(release, 'diff_cover', 'python3-diff-cover', '--no-build --no-init -p 3')
         },
         integration: {
-            if(release == 'f29') {
-                bodhi_ci(release, 'integration', 'python3-integration', '--no-build --no-init -p 3')
-            }
+            bodhi_ci(release, 'integration', 'python3-integration', '--no-init -p 3')
         }
     )
 }

--- a/devel/ci/integration/bodhi/Dockerfile-f28
+++ b/devel/ci/integration/bodhi/Dockerfile-f28
@@ -1,0 +1,55 @@
+FROM bodhi-ci/f28
+LABEL \
+  name="bodhi-web" \
+  vendor="Fedora Infrastructure" \
+  maintainer="Aurelien Bompard <abompard@fedoraproject.org>" \
+  license="MIT"
+
+# For integration testing we're using the infrastructure repo
+RUN curl -o /etc/yum.repos.d/infra-tags.repo https://infrastructure.fedoraproject.org/cgit/ansible.git/plain/files/common/fedora-infra-tags.repo
+RUN dnf upgrade -y
+
+# Install Bodhi deps (that were not needed by the unittests container)
+RUN dnf install -y \
+    httpd \
+    intltool \
+    python3-mod_wsgi \
+    "python3dist(psycopg2)" \
+    skopeo \
+    /usr/bin/koji
+
+# Create bodhi user
+RUN groupadd -r bodhi && \
+    useradd  -r -s /sbin/nologin -d /home/bodhi/ -m -c 'Bodhi Server' -g bodhi bodhi
+
+# Install it
+RUN python3 setup.py build && pip3 install .
+
+# Configuration
+RUN mkdir -p /etc/bodhi
+COPY production.ini /etc/bodhi/production.ini
+
+COPY devel/ci/integration/bodhi/start.sh /etc/bodhi/start.sh
+COPY devel/ci/integration/bodhi/httpd.conf /etc/bodhi/httpd.conf
+COPY apache/bodhi.wsgi /etc/bodhi/bodhi.wsgi
+RUN sed -i -e 's,/var/www,/httpdir,g' /etc/bodhi/bodhi.wsgi
+# Put the fedmsg config in place. Let's name it starting with zz so it sorts last.
+COPY devel/ci/integration/bodhi/fedmsg.py /etc/fedmsg.d/zzbodhi.py
+
+RUN \
+# Put the fedmsg cert and key in place
+    mkdir -p /etc/pki/fedmsg/ && \
+    ln -sf /etc/pki/fedmsg/key/fedmsg-bodhi.key /etc/pki/fedmsg/bodhi.key && \
+    ln -sf /etc/pki/fedmsg/crt/fedmsg-bodhi.crt /etc/pki/fedmsg/bodhi.crt && \
+# Set up krb5
+    rm -f /etc/krb5.conf && \
+    ln -sf /etc/bodhi/krb5.conf /etc/krb5.conf && \
+    ln -sf /etc/keytabs/koji-keytab /etc/krb5.bodhi_bodhi.fedoraproject.org.keytab
+
+# Apache
+RUN mkdir -p /httpdir && chown bodhi:bodhi /httpdir
+
+EXPOSE 8080
+USER bodhi
+ENV USER=bodhi
+CMD ["bash", "/etc/bodhi/start.sh"]

--- a/devel/ci/integration/bodhi/Dockerfile-f29
+++ b/devel/ci/integration/bodhi/Dockerfile-f29
@@ -1,82 +1,28 @@
-FROM registry.fedoraproject.org/fedora:29
+FROM bodhi-ci/f29
 LABEL \
   name="bodhi-web" \
   vendor="Fedora Infrastructure" \
   maintainer="Aurelien Bompard <abompard@fedoraproject.org>" \
   license="MIT"
 
+# For integration testing we're using the infrastructure repo
 RUN curl -o /etc/yum.repos.d/infra-tags.repo https://infrastructure.fedoraproject.org/cgit/ansible.git/plain/files/common/fedora-infra-tags.repo
+RUN dnf upgrade -y
 
-# While dnf has a --nodocs, it doesen't have a --docs...
-RUN sed -i '/nodocs/d' /etc/dnf/dnf.conf
-
-# Install Bodhi deps
-# 1. bodhi-server
-# 2. bodhi-composer
-# 3. python3-bodhi-client
+# Install Bodhi deps (that were not needed by the unittests container)
 RUN dnf install -y \
-    \
-    fedora-messaging \
     httpd \
     intltool \
-    liberation-mono-fonts \
-    python3-dogpile-cache \
-    python3-koji \
-    python3-librepo \
     python3-mod_wsgi \
-    python3-pillow \
-    "python3dist(alembic)" \
-    "python3dist(arrow)" \
-    "python3dist(bleach)" \
-    "python3dist(click)" \
-    "python3dist(colander)" \
-    "python3dist(cornice)" \
-    "python3dist(cryptography)" \
-    "python3dist(fedmsg)" \
-    "python3dist(feedgen)" \
-    "python3dist(jinja2)" \
-    "python3dist(kitchen)" \
-    "python3dist(markdown)" \
     "python3dist(psycopg2)" \
-    "python3dist(pylibravatar)" \
-    "python3dist(pyramid)" \
-    "python3dist(pyramid-fas-openid)" \
-    "python3dist(pyramid-mako)" \
-    "python3dist(python-bugzilla)" \
-    "python3dist(python-fedora)" \
-    "python3dist(python3-openid)" \
-    "python3dist(simplemediawiki)" \
-    "python3dist(six)" \
-    "python3dist(sqlalchemy)" \
-    "python3dist(waitress)" \
-    \
-    python3-createrepo_c \
-    python3-hawkey \
     skopeo \
-    \
-    /usr/bin/koji \
-    python3-dnf \
-    python3-koji \
-    "python3dist(click)" \
-    "python3dist(iniparse)" \
-    "python3dist(kitchen)" \
-    "python3dist(python-fedora)" \
-    "python3dist(six)" \
-    \
-    fedmsg \
-    fedmsg-base \
-    git
+    /usr/bin/koji
 
 # Create bodhi user
 RUN groupadd -r bodhi && \
     useradd  -r -s /sbin/nologin -d /home/bodhi/ -m -c 'Bodhi Server' -g bodhi bodhi
 
-# Copy source
-WORKDIR /bodhi
-COPY . /bodhi
-
 # Install it
-RUN sed -i -e '/pyramid_debugtoolbar/d' setup.py devel/development.ini.example
 RUN python3 setup.py build && pip3 install .
 
 # Configuration

--- a/devel/ci/integration/bodhi/Dockerfile-pip
+++ b/devel/ci/integration/bodhi/Dockerfile-pip
@@ -1,0 +1,62 @@
+FROM bodhi-ci/pip
+LABEL \
+  name="bodhi-web" \
+  vendor="Fedora Infrastructure" \
+  maintainer="Aurelien Bompard <abompard@fedoraproject.org>" \
+  license="MIT"
+
+# For integration testing we're using the infrastructure repo
+RUN curl -o /etc/yum.repos.d/infra-tags.repo https://infrastructure.fedoraproject.org/cgit/ansible.git/plain/files/common/fedora-infra-tags.repo
+RUN dnf upgrade -y
+
+# Install Bodhi deps (that were not needed by the unittests container)
+RUN dnf install -y \
+    httpd \
+    intltool \
+    python3-koji \
+    /usr/bin/koji \
+    python3-mod_wsgi \
+    python3-dnf \
+    skopeo
+
+RUN pip-3 install \
+    psycopg2
+
+# Mimic RPM's handling of Python3-specific binaries
+RUN ln -s alembic /usr/local/bin/alembic-3
+
+# Create bodhi user
+RUN groupadd -r bodhi && \
+    useradd  -r -s /sbin/nologin -d /home/bodhi/ -m -c 'Bodhi Server' -g bodhi bodhi
+
+# Install it
+RUN python3 setup.py build && pip3 install .
+
+# Configuration
+RUN mkdir -p /etc/bodhi
+COPY production.ini /etc/bodhi/production.ini
+
+COPY devel/ci/integration/bodhi/start.sh /etc/bodhi/start.sh
+COPY devel/ci/integration/bodhi/httpd.conf /etc/bodhi/httpd.conf
+COPY apache/bodhi.wsgi /etc/bodhi/bodhi.wsgi
+RUN sed -i -e 's,/var/www,/httpdir,g' /etc/bodhi/bodhi.wsgi
+# Put the fedmsg config in place. Let's name it starting with zz so it sorts last.
+COPY devel/ci/integration/bodhi/fedmsg.py /etc/fedmsg.d/zzbodhi.py
+
+RUN \
+# Put the fedmsg cert and key in place
+    mkdir -p /etc/pki/fedmsg/ && \
+    ln -sf /etc/pki/fedmsg/key/fedmsg-bodhi.key /etc/pki/fedmsg/bodhi.key && \
+    ln -sf /etc/pki/fedmsg/crt/fedmsg-bodhi.crt /etc/pki/fedmsg/bodhi.crt && \
+# Set up krb5
+    rm -f /etc/krb5.conf && \
+    ln -sf /etc/bodhi/krb5.conf /etc/krb5.conf && \
+    ln -sf /etc/keytabs/koji-keytab /etc/krb5.bodhi_bodhi.fedoraproject.org.keytab
+
+# Apache
+RUN mkdir -p /httpdir && chown bodhi:bodhi /httpdir
+
+EXPOSE 8080
+USER bodhi
+ENV USER=bodhi
+CMD ["bash", "/etc/bodhi/start.sh"]

--- a/devel/ci/integration/bodhi/Dockerfile-rawhide
+++ b/devel/ci/integration/bodhi/Dockerfile-rawhide
@@ -1,0 +1,55 @@
+FROM bodhi-ci/rawhide
+LABEL \
+  name="bodhi-web" \
+  vendor="Fedora Infrastructure" \
+  maintainer="Aurelien Bompard <abompard@fedoraproject.org>" \
+  license="MIT"
+
+# For integration testing we're using the infrastructure repo
+RUN curl -o /etc/yum.repos.d/infra-tags.repo https://infrastructure.fedoraproject.org/cgit/ansible.git/plain/files/common/fedora-infra-tags.repo
+RUN dnf upgrade -y
+
+# Install Bodhi deps (that were not needed by the unittests container)
+RUN dnf install -y \
+    httpd \
+    intltool \
+    python3-mod_wsgi \
+    "python3dist(psycopg2)" \
+    skopeo \
+    /usr/bin/koji
+
+# Create bodhi user
+RUN groupadd -r bodhi && \
+    useradd  -r -s /sbin/nologin -d /home/bodhi/ -m -c 'Bodhi Server' -g bodhi bodhi
+
+# Install it
+RUN python3 setup.py build && pip3 install .
+
+# Configuration
+RUN mkdir -p /etc/bodhi
+COPY production.ini /etc/bodhi/production.ini
+
+COPY devel/ci/integration/bodhi/start.sh /etc/bodhi/start.sh
+COPY devel/ci/integration/bodhi/httpd.conf /etc/bodhi/httpd.conf
+COPY apache/bodhi.wsgi /etc/bodhi/bodhi.wsgi
+RUN sed -i -e 's,/var/www,/httpdir,g' /etc/bodhi/bodhi.wsgi
+# Put the fedmsg config in place. Let's name it starting with zz so it sorts last.
+COPY devel/ci/integration/bodhi/fedmsg.py /etc/fedmsg.d/zzbodhi.py
+
+RUN \
+# Put the fedmsg cert and key in place
+    mkdir -p /etc/pki/fedmsg/ && \
+    ln -sf /etc/pki/fedmsg/key/fedmsg-bodhi.key /etc/pki/fedmsg/bodhi.key && \
+    ln -sf /etc/pki/fedmsg/crt/fedmsg-bodhi.crt /etc/pki/fedmsg/bodhi.crt && \
+# Set up krb5
+    rm -f /etc/krb5.conf && \
+    ln -sf /etc/bodhi/krb5.conf /etc/krb5.conf && \
+    ln -sf /etc/keytabs/koji-keytab /etc/krb5.bodhi_bodhi.fedoraproject.org.keytab
+
+# Apache
+RUN mkdir -p /httpdir && chown bodhi:bodhi /httpdir
+
+EXPOSE 8080
+USER bodhi
+ENV USER=bodhi
+CMD ["bash", "/etc/bodhi/start.sh"]

--- a/devel/ci/integration/tests/fixtures/backend.py
+++ b/devel/ci/integration/tests/fixtures/backend.py
@@ -44,7 +44,7 @@ def docker_network(docker_backend):
         conu.DockerBackend: The Docker backend fixture.
 
     Yields:
-        str: The Docker network ID.
+        dict: The Docker network.
     """
     network = docker_backend.d.create_network("bodhi_test", driver="bridge")
     yield network

--- a/devel/ci/integration/tests/fixtures/bodhi.py
+++ b/devel/ci/integration/tests/fixtures/bodhi.py
@@ -20,7 +20,6 @@ import os
 from uuid import uuid4
 
 import pytest
-from conu.backend.docker.container_parameters import DockerContainerParameters
 from six.moves.configparser import ConfigParser
 
 from ..utils import make_db_and_user, edit_file
@@ -55,7 +54,7 @@ def bodhi_container(
     image = docker_backend.ImageClass(
         os.environ.get("BODHI_INTEGRATION_IMAGE", "bodhi-ci-integration-bodhi")
     )
-    container = image.run_via_api(DockerContainerParameters(name="bodhi"))
+    container = image.run_via_api()
     config_override = {
         "base_address": "https://bodhi.fedoraproject.org/",
         "sqlalchemy.url": "postgresql://bodhi2@db/bodhi2",
@@ -116,7 +115,9 @@ def bodhi_container(
             config.write(config_file)
 
     container.start()
-    docker_backend.d.connect_container_to_network(container.name, docker_network["Id"])
+    docker_backend.d.connect_container_to_network(
+        container.get_id(), docker_network["Id"], aliases=["bodhi"],
+    )
     # Update the database schema
     container.execute(["alembic-3", "-c", "/bodhi/alembic.ini", "upgrade", "head"])
     # we need to wait for the webserver to start serving

--- a/devel/ci/integration/tests/fixtures/db.py
+++ b/devel/ci/integration/tests/fixtures/db.py
@@ -18,7 +18,6 @@
 
 import os
 import pytest
-from conu.backend.docker.container_parameters import DockerContainerParameters
 
 
 @pytest.fixture(scope="session")
@@ -38,11 +37,11 @@ def db_container(docker_backend, docker_network):
         os.environ.get("BODHI_INTEGRATION_POSTGRESQL_IMAGE", "postgres"),
         tag="latest"
     )
-    container = image.run_via_api(
-        DockerContainerParameters(name="db")
-    )
+    container = image.run_via_api()
     container.start()
-    docker_backend.d.connect_container_to_network(container.name, docker_network["Id"])
+    docker_backend.d.connect_container_to_network(
+        container.get_id(), docker_network["Id"], aliases=["db"],
+    )
     container.wait_for_port(5432, timeout=-1)
     yield container
     container.kill()

--- a/devel/ci/integration/tests/fixtures/greenwave.py
+++ b/devel/ci/integration/tests/fixtures/greenwave.py
@@ -17,7 +17,6 @@
 # Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 
 import pytest
-from conu.backend.docker.container_parameters import DockerContainerParameters
 
 from ..utils import make_db_and_user
 
@@ -39,11 +38,11 @@ def greenwave_container(docker_backend, docker_network, db_container):
     # Define the container and start it
     image_name = "bodhi-ci-integration-greenwave"
     image = docker_backend.ImageClass(image_name)
-    container = image.run_via_api(
-        DockerContainerParameters(name="greenwave")
-    )
+    container = image.run_via_api()
     container.start()
-    docker_backend.d.connect_container_to_network(container.name, docker_network["Id"])
+    docker_backend.d.connect_container_to_network(
+        container.get_id(), docker_network["Id"], aliases=["greenwave"],
+    )
     # we need to wait for the webserver to start serving
     container.wait_for_port(8080, timeout=-1)
     yield container

--- a/devel/ci/integration/tests/fixtures/resultsdb.py
+++ b/devel/ci/integration/tests/fixtures/resultsdb.py
@@ -17,7 +17,6 @@
 # Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 
 import pytest
-from conu.backend.docker.container_parameters import DockerContainerParameters
 
 from ..utils import make_db_and_user
 
@@ -39,11 +38,11 @@ def resultsdb_container(docker_backend, docker_network, db_container):
     # Define the container and start it
     image_name = "bodhi-ci-integration-resultsdb"
     image = docker_backend.ImageClass(image_name)
-    container = image.run_via_api(
-        DockerContainerParameters(name="resultsdb")
-    )
+    container = image.run_via_api()
     container.start()
-    docker_backend.d.connect_container_to_network(container.name, docker_network["Id"])
+    docker_backend.d.connect_container_to_network(
+        container.get_id(), docker_network["Id"], aliases=["resultsdb"],
+    )
     # Add sample data in the database
     container.execute(["resultsdb", "init_db"])
     # we need to wait for the webserver to start serving

--- a/devel/ci/integration/tests/fixtures/waiverdb.py
+++ b/devel/ci/integration/tests/fixtures/waiverdb.py
@@ -17,7 +17,6 @@
 # Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 
 import pytest
-from conu.backend.docker.container_parameters import DockerContainerParameters
 
 from ..utils import make_db_and_user
 
@@ -42,11 +41,11 @@ def waiverdb_container(docker_backend, docker_network, db_container):
     # Define the container and start it
     image_name = "bodhi-ci-integration-waiverdb"
     image = docker_backend.ImageClass(image_name)
-    container = image.run_via_api(
-        DockerContainerParameters(name="waiverdb")
-    )
+    container = image.run_via_api()
     container.start()
-    docker_backend.d.connect_container_to_network(container.name, docker_network["Id"])
+    docker_backend.d.connect_container_to_network(
+        container.get_id(), docker_network["Id"], aliases=["waiverdb"],
+    )
     # Add sample data in the database
     container.execute(["waiverdb", "db", "upgrade"])
     # we need to wait for the webserver to start serving

--- a/devel/ci/integration/tests/test_bodhi_cli.py
+++ b/devel/ci/integration/tests/test_bodhi_cli.py
@@ -21,8 +21,9 @@ import json
 import re
 import textwrap
 
-import psycopg2
 import requests
+import psycopg2
+import pytest
 
 from conu import ConuException
 from munch import Munch
@@ -154,7 +155,7 @@ def test_overrides_query(bodhi_container, db_container):
     result = _run_cli(bodhi_container, ["overrides", "query"])
     assert result.exit_code == 0
     last_line = result.output.split("\n")[-2]
-    assert last_line == "{} overrides found (20 shown)".format(total)
+    assert last_line == "{} overrides found ({} shown)".format(total, min(total, 20))
 
 
 def test_updates_query_total(bodhi_container, db_container):
@@ -172,7 +173,7 @@ def test_updates_query_total(bodhi_container, db_container):
     result = _run_cli(bodhi_container, ["updates", "query"])
     assert result.exit_code == 0
     last_line = result.output.split("\n")[-2]
-    assert last_line == "{} updates found (20 shown)".format(total)
+    assert last_line == "{} updates found ({} shown)".format(total, min(total, 20))
 
 
 def test_updates_query_details(bodhi_container, db_container, greenwave_container):
@@ -209,7 +210,10 @@ def test_updates_query_details(bodhi_container, db_container, greenwave_containe
     with conn:
         with conn.cursor() as curs:
             curs.execute(query_update)
-            update = _db_record_to_munch(curs, curs.fetchone())
+            result = curs.fetchone()
+            if result is None:
+                pytest.skip("No update in the database")
+            update = _db_record_to_munch(curs, result)
             update.comments = []
             curs.execute(query_comments, (update.id, ))
             for record in curs:

